### PR TITLE
Extract build toc document step base

### PR DESCRIPTION
**Pull request by hellosnow**

### CODE REVIEW

File: src/Microsoft.DocAsCode.Build.TableOfContents/TocRestruction.cs [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+namespace Microsoft.DocAsCode.Build.TableOfContents
+{
+    using System;
+    using System.Collections.Generic;
+
+    using Microsoft.DocAsCode.Common;
+    using Microsoft.DocAsCode.DataContracts.Common;
+    using Microsoft.DocAsCode.Plugins;
+
+    public static class TocRestruction
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.TableOfContents/TocRestruction.cs]: >public static class TocRestruction [](start = 4, length = 34)

internal static class TocRestructureUtility?

File: src/Microsoft.DocAsCode.Build.TableOfContents/BuildTocDocumentStepBase.cs [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    using System.Linq;
+
+    using Microsoft.DocAsCode.Build.Common;
+    using Microsoft.DocAsCode.Common;
+    using Microsoft.DocAsCode.DataContracts.Common;
+    using Microsoft.DocAsCode.Plugins;
+
+    public abstract class BuildTocDocumentStepBase : BaseDocumentBuildStep
+    {
+        internal ImmutableDictionary<string, TocItemInfo> TocModelCache = ImmutableDictionary<string, TocItemInfo>.Empty;
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.TableOfContents/BuildTocDocumentStepBase.cs]: >TocModelCache [](start = 58, length = 13)

it makes the build step not thread safe, get rid of instance level variables?

Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.TableOfContents/BuildTocDocumentStepBase.cs]: how about pass it via parameter?

---
In reply to: [121076786](https://github.com/dotnet/docfx/pull/1750#discussion_r121076786) [](ancestors = 121076786)

File: src/Microsoft.DocAsCode.Build.TableOfContents/TocItemInfo.cs [hunk 3]
```diff
@@ -6,7 +6,7 @@ namespace Microsoft.DocAsCode.Build.TableOfContents
     using Microsoft.DocAsCode.DataContracts.Common;
     using Microsoft.DocAsCode.Plugins;
 
-    internal sealed class TocItemInfo
+    public sealed class TocItemInfo
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.TableOfContents/TocItemInfo.cs]: >public [](start = 4, length = 6)

why?

File: src/Microsoft.DocAsCode.Build.TableOfContents/BuildTocDocumentStepBase.cs [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+            ReportBuildDependency(model, host);
+            model.Content = toc;
+            // todo : metadata.
+        }
+
+        #endregion
+
+        #region Abstract methods
+
+        public abstract void ReportBuildDependency(FileModel model, IHostService host);
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.TableOfContents/BuildTocDocumentStepBase.cs]: >ReportBuildDependency [](start = 29, length = 21)

is the method name too specific?

File: src/Microsoft.DocAsCode.Build.TableOfContents/TocItemInfo.cs (hunk 3 above)
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.TableOfContents/TocItemInfo.cs]: reason is TocResolverUtility.Resolve is using it and be public. Plan to change the return type to TocItemViewModel in next iteration

---
In reply to: [121313790](https://github.com/dotnet/docfx/pull/1750#discussion_r121313790) [](ancestors = 121313790)

File: src/Microsoft.DocAsCode.Build.TableOfContents/BuildTocDocumentStepBase.cs (hunk 4 above)
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.TableOfContents/BuildTocDocumentStepBase.cs]: or "BuildPostActions"? Any suggestions?

---
In reply to: [121313850](https://github.com/dotnet/docfx/pull/1750#discussion_r121313850) [](ancestors = 121313850)

File: src/Microsoft.DocAsCode.Build.TableOfContents/TocResolverUtility.cs [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+            foreach (var key in tocCache.Keys.ToList())
+            {
+                tocCache[key] = tocResolver.Resolve(key);
+            }
+
+            foreach (var model in models)
+            {
+                if (!tocCache.TryGetValue(model.OriginalFileAndType.FullPath, out TocItemInfo tocItemInfo))
+                {
+                    throw new DocfxException($"File {model.OriginalFileAndType.FullPath} cannot be found from toc cache.");
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.TableOfContents/TocResolverUtility.cs]: >throw new DocfxException($"File {model.OriginalFileAndType.FullPath} cannot be found from toc cache."); [](start = 20, length = 103)

never happens according to L24?

File: src/Microsoft.DocAsCode.Build.TableOfContents/TocResolverUtility.cs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    using Microsoft.DocAsCode.Exceptions;
+
+    public class TocResolverUtility
+    {
+        public static IEnumerable<FileModel> Resolve(ImmutableList<FileModel> models, IHostService host)
+        {
+            var tocCache = new Dictionary<string, TocItemInfo>(FilePathComparer.OSPlatformSensitiveStringComparer);
+            foreach (var model in models)
+            {
+                if (!tocCache.ContainsKey(model.OriginalFileAndType.FullPath))
```
Comment (vicancy) [inline src/Microsoft.DocAsCode.Build.TableOfContents/TocResolverUtility.cs]: >if (!tocCache.ContainsKey(model.OriginalFileAndType.FullPath)) [](start = 16, length = 62)

no need to check?

**Comment by vicancy** [comment]
:shipit:

**Pull request merged**